stl/inc/vector

```diff
@@ -440,6 +440,189 @@ constexpr _Ty* _Unfancy_maybe_null(_Ty* _Ptr) noexcept { // do nothing for plain
+    : bool_constant<_Vec::allocator_type::_Minimum_allocation_alignment >= 8> {};
+
+template <class _Ty, bool = _Has_minimum_allocation_alignment<_Ty>::value>
+class _NODISCARD _Asan_annotater {
+private:
+    using _Alloc      = typename _Ty::allocator_type;
+    using _Pointer    = typename _Ty::pointer;
+    using _Size_type  = typename _Ty::size_type;
+    using _Value_type = typename _Ty::value_type;
+
+public:
+    constexpr _Asan_annotater(const _Ty& _Target) : _Target_vec(_Target), _Old_last(nullptr) {}
+
+    _CONSTEXPR20 _Asan_annotater(const _Ty& _Target, _Size_type _New_size) : _Target_vec(_Target) {
+#ifdef __cpp_lib_constexpr_dynamic_alloc
+        if (_STD is_constant_evaluated()) {
+            return;
+        }
+#endif // __cpp_lib_constexpr_dynamic_alloc
+
+        const auto& _My_data     = _Target_vec._Mypair._Myval2;
+        const _Pointer& _Myfirst = _My_data._Myfirst;
+        const _Pointer& _Mylast  = _My_data._Mylast;
+        const _Pointer& _Myend   = _My_data._Myend;
+
+        _Size_type _Old_capacity = static_cast<_Size_type>(_Myend - _Myfirst);
+
+        if (_New_size > _Old_capacity) {
+            _Old_last = nullptr;
+            return;
+        }
+
+        _Old_last            = _Mylast;
+        _Size_type _Old_size = static_cast<_Size_type>(_Mylast - _Myfirst);
+
+        if (_New_size > _Old_size) {
+            __sanitizer_annotate_contiguous_container(_Myfirst, _Myend, _Old_last, _Myfirst + _New_size);
+        }
+    }
+
+    _CONSTEXPR20 ~_Asan_annotater() {
+#ifdef __cpp_lib_constexpr_dynamic_alloc
+        if (_STD is_constant_evaluated()) {
+            return;
+        }
+#endif // __cpp_lib_constexpr_dynamic_alloc
+
+        const auto& _My_data     = _Target_vec._Mypair._Myval2;
+        const _Pointer& _Myfirst = _My_data._Myfirst;
+        const _Pointer& _Mylast  = _My_data._Mylast;
+        const _Pointer& _Myend   = _My_data._Myend;
+
+        if (!_Old_last) {
+            __sanitizer_annotate_contiguous_container(_Myfirst, _Myend, _Myend, _Mylast);
+        } else if (_Mylast < _Old_last) {
+            __sanitizer_annotate_contiguous_container(_Myfirst, _Myend, _Old_last, _Mylast);
+        }
+    }
+
+private:
+    const _Ty& _Target_vec;
+
+    _Pointer _Old_last;
+};
+
+template <class _Ty>
+class _NODISCARD _Asan_annotater<_Ty, false> {
+private:
+    using _Alloc      = typename _Ty::allocator_type;
+    using _Pointer    = typename _Ty::pointer;
+    using _Size_type  = typename _Ty::size_type;
+    using _Value_type = typename _Ty::value_type;
+
+    const void* _Get_aligned_start() {
+        const auto& _My_data       = _Target_vec._Mypair._Myval2;
+        const _Pointer& _Myfirst   = _My_data._Myfirst;
+        const _Pointer& _Myend     = _My_data._Myend;
+        const _Size_type _Capacity = static_cast<_Size_type>(_Myend - _Myfirst);
+
+        if (_Capacity * sizeof(_Value_type) < 8) {

```

Comment:
could we invert and make it an early return after the `reinterpret_cast`?

Reply:
I'm not exactly sure what you want me to invert here. Do the reinterpret cast on the next line first before the check for `< 8`?

Reply:
I thought about inverting the condition so that you can return after the single statement in the other branch